pkg/api/query/v1.go

```diff
@@ -1301,6 +1311,54 @@ func (qapi *QueryAPI) stores(_ *http.Request) (interface{}, []error, *api.ApiErr
 	return statuses, nil, nil, func() {}
 }
 
+func (qapi *QueryAPI) getLabelMatchers(matchers []string, tenant string) ([][]*labels.Matcher, *api.ApiError) {
+	tenantLabelMatcher := &labels.Matcher{
+		Name:  qapi.tenantLabel,
+		Type:  labels.MatchEqual,
+		Value: tenant,
+	}
+
+	matcherSets := make([][]*labels.Matcher, 0, len(matchers))
+
+	// If tenancy is enforced, but there are no matchers at all, add the tenant matcher
+	if len(matchers) == 0 && qapi.enforceTenancy {
+		var matcher []*labels.Matcher
+		matcher = append(matcher, tenantLabelMatcher)
+		matcherSets = append(matcherSets, matcher)
+		return matcherSets, nil
+	}
+
+	for _, s := range matchers {
+		matchers, err := parser.ParseMetricSelector(s)
+		if err != nil {
+			return nil, &api.ApiError{Typ: api.ErrorBadData, Err: err}
+		}
+		if qapi.enforceTenancy {
```

Comment:
We can probably flip this and return early, and do the changes otherwise.
Or even check this in the query/queryRange functions themselves.
```suggestion
		if !qapi.enforceTenancy {
			matcherSets = append(matcherSets, matchers)
			continue
		}
```

Reply:
The appending of `matchers` to `matcherSets` happen in both cases, and needs to happen after the tenancy specific logic. So it for me it made more sense to me like this, otherwise we'd need to do the appending twice, i.e:

```
if !qapi.enforceTenancy {
	matcherSets = append(matcherSets, matchers)
	continue
}
... tenancy specific logic to change matcher
matcherSets = append(matcherSets, matchers)
```

If you still think it's better like that, I'm happy to do the change. Just making sure you've noticed the append then need be written twice...

Reply:
Ah yes, noticed that twice append, but wanted to exit early and move the code left, instead of keeping everything as an if. But I'm fine with this too, so not a blocker

Reply:
I left is "as is", but with using prom-label-proxys functionality the `if` statement becomes quite small, so hopefully that makes it better overall.